v0/destinations/kustomer/transform.js

```diff
@@ -0,0 +1,226 @@
+    kustomerPayload.name = `${firstName} ${lastName}`;
+  }
+
+  if (email) {
+    kustomerPayload.emails = kustomerPayload.emails
+      ? kustomerPayload.emails
+      : [{ type: "home", email }];
+  }
+
+  if (phone) {
+    kustomerPayload.phones = kustomerPayload.phones
+      ? kustomerPayload.phones
+      : [{ type: "home", phone }];
+  }
+  if (url) {
+    kustomerPayload.urls = [
+      {
+        url
+      }
+    ];
+  }
+
+  const address = getFieldValueFromMessage(message, "address");
+  if (address) {
+    let addrStr;
+    if (typeof address === "string") {
+      addrStr = address;
+    } else {
+      const { street, city, state, postalCode } = address;
+      addrStr =
+        street || city || state || postalCode
+          ? `${street || ""} ${city || ""} ${state || ""} ${postalCode || ""}`
+          : addrStr;
+    }
+    if (typeof addrStr === "string") {
+      kustomerPayload.locations = [
+        {
+          type: "home",
+          address: addrStr
+        }
+      ];
+    }
+  }
+
+  removeUndefinedAndNullValues(kustomerPayload);
+  return kustomerPayload;
+};
+
+// Main process function responsible for building payload for all
+// type of events.
+const responseBuilderSimple = async (message, category, destination) => {
+  let payload = {};
+  let targetUrl;
+  let storedState = {
+    userExists: false
+  };
+  // In case of identify type of event first extract the anonymousId, userId
+  // and search if any Kustomer is present in destination.
+  // If present update the same kustomer with the given payload.
+  // else create a new kustomer.
+  //-------------------------------------------------------------
+  // Get Kustomer: https://apidocs.kustomer.com/#ff41f372-6144-4c64-9712-662ee5ef1c33
+  // Create Kustomer: https://apidocs.kustomer.com/#07bd1072-4d4b-4875-b526-8369d711e811
+  // Update Kustomer: https://apidocs.kustomer.com/#077d653a-184e-4153-8133-d24b6427c1ae
+  if (message.type.toLowerCase() == EventType.IDENTIFY) {
+    const userEmail = getFieldValueFromMessage(message, "email");
+    const userId = getFieldValueFromMessage(message, "userIdOnly");
+    const anonymousId = get(message, "anonymousId");
+    const externalId = getDestinationExternalID(message, "kustomerExternalId");
+    if (externalId) {
+      storedState = {
+        userExists: true,
+        targetUrl: `${BASE_ENDPOINT}/v1/customers/${externalId}?replace=false`
+      };
+    }
+    // If email exists we first search Kustomer with email if present then we mark it
+    // for update call.
+    if (!storedState.userExists && userEmail) {
+      storedState = await fetchKustomer(
+        `${BASE_ENDPOINT}/v1/customers/email=${userEmail}`,
+        destination
+      );
+    }
+    // If response.userExists flag is false
+    // If Kustomer has userId we search using userId as externalId if user
+    // is present or not. If yes then we mark it for update.
+    if (!storedState.userExists && userId) {
```

Comment:
if stored state gets a value in the previous if should it go to this if statement?

Reply:
```
let storedState = {
    userExists: false
  };
  
  ```
  initially, the `storedsState` with `userExists` property as false
  if we have externalId we update the `storedState` as -->
  
 ```
  if (externalId) {
      storedState = {
        userExists: true,
        targetUrl: `${BASE_ENDPOINT}/v1/customers/${externalId}?replace=false`
      };
    }
    
```
   
   Else we do  lookup in the destination using `email, userId and anonymousId` in that order
    
   wehenever we get a 200 response and a response object meaning we have stored customer against that `email,userid, or anonymousId`
   We mark ther `storedState` as
   
   ```
    if (response && response.status === 200 && response.data.data.id) {
    return {
      userExists: true,
      targetUrl: `${BASE_ENDPOINT}/v1/customers/${response.data.data.id}?replace=false`
    };
  }
  
  ```
  else
  
  ```
  return { userExists: false };
  ```
  
  Finally
  ```
  // URL to use for creating new Kustomer
    if (!storedState.userExists) {
      targetUrl = `${BASE_ENDPOINT}/v1/customers`;
    }
    // URL to use for updating Kustomer
    else {
      targetUrl = storedState.targetUrl;
    }
  ```

Reply:
got it..